profile static view

matallo opened the pull request:
migrate `/profile` view from Rails to static view

- [x] add webpack config https://github.com/CartoDB/cartodb/issues/12674
- [x] migrate rails to js https://github.com/CartoDB/cartodb/issues/12668
- [x] implement forms https://github.com/CartoDB/cartodb/issues/12669
- [x] add avatar valid extensions https://github.com/CartoDB/cartodb/issues/12771
- [x] add new api endpoints for user account deletion https://github.com/CartoDB/cartodb/issues/12841

~~this will eventually point to `master` once https://github.com/CartoDB/cartodb/pull/12680 is merged~~

---

when setting the `static_profile` FeatureFlag to a user, the profile will load from a static html

- as "implement forms" issue involved creating new tests (for the new functionality), it has been restored the possibility to run the tests in the browser

- there are some fields added tot the user presenter, and also the avatar helper require, maybe worth taking a look by @CartoDB/backend 

https://github.com/CartoDB/cartodb/pull/12704/files#diff-c6081688a012bc9178036ba7f2153227R193

- there are some tests which were missing from the static dashboard, we should have the new views covered now, mostly wrote them from scratch following unit testing fashion but `content_controller_view_static.spec.js` which is mostly to ensure the behaviour will be the same

- duplicated the `flash_message` view from organization to be able to pass type (and not mess up with current existing views), if we finally migrate this to Builder these should be consolidated

- moved the `vendor_scripts` view from `dashboard` to `common` now it is a shared view, and fixed a bug with the variables surfaced in the specs 👼 

- during the templates migration I spotted some issue with the footer view, where an onpremises email address is shown to self hosted users 🤷‍♂️ 

https://github.com/CartoDB/cartodb/blob/master/lib/assets/javascripts/cartodb/common/views/footer_static.jst.ejs#L14 => https://github.com/CartoDB/cartodb/pull/12704/files#diff-617908fc0c95f1b609ddd1e7c17f1edbR13

- I'm not pretty sure how to test the entry js points

---

**acceptance**

- submit form and update user successfully
- check `flash_message` messages when form submission fails

matallo commented:
currently the js assets for profile are reused with `account` with stuff like `if (this.$('.js-deleteAccount').length > 0 && window.authenticity_token) {` to target account view elements.

I'm splitting account and profile assets, migrating the rails view to a js template for profile, but in the future we could effectively serve the same html and assets, and add a router so it behaves similarly to maps/datasets in the dashboard

maybe this can be tackled after as a separate task, and improve perceived performance

matallo commented:
actually this view doesn't need validation, so as soon as the new users API is ready this would be too

juanignaciosl commented:
Ok, I'm CR-ing this, but the checklist on top still has one pending task: "implement forms #12669".

juanignaciosl commented:
_EDIT: frontend tests are failing, back._

matallo commented:
just for the record, this PR is not ready until we have the new users api

matallo commented:
tests pending

matallo commented:
restored tests in the browser, now back to writing the actual specs

matallo commented:
finished the specs, ready for review!

matallo commented:
retest this, please

nobuti commented on CONTRIBUTING.md:
```diff
@@ -107,7 +107,17 @@ grunt editor_specs
 After the building process finish, a webpage will show up with a link to the Jasmine page with all the specs. The URL of this page is `http://localhost:8089/_SpecRunner.html`
 
-Editor specs process won't watch for changes, so every time you modify any file you'll have to run again the Editor specs command.
+Then, the process will watch changes in the codebase and will regenerate the specs as needed. Just refresh the Jasmine page to pass again the tests.
```
Thanks 🙌

nobuti commented on lib/assets/javascripts/cartodb/account/views/private_header.jst.ejs:
```diff
@@ -0,0 +1,36 @@
+      <li>
+        <ul class="Header-navigationBreadcrumb">
+          <% if (organizationName) { %>
+            <li class="Header-navigationBreadcrumbItem CDB-Text CDB-Size-large"><p class="Header-navigationBreadcrumbParagraph"><%= organizationName %></p> /</li>
```
Maybe we could fix those `/` from style, as pseudo-elements. It's a bit hard to deal with spaces and so.

matallo replied:
At this point I only focused on the migration, if we think this is something to improve we should open new issues, but happy to see suggestions!

nobuti commented on lib/assets/javascripts/cartodb/dashboard/main_view.js:
```diff
@@ -127,50 +128,56 @@ module.exports = cdb.core.View.extend({
     });
-
-    supportView.render();
+    this.$el.append(vendorScriptsView.render().el);
```
no `this.addView(vendorScriptsView)` for these lines?

matallo replied:
actually I don't know how I confused this, this file should remain untouched, the changes should be at https://github.com/CartoDB/cartodb/blob/12674-static-profile/lib/assets/javascripts/cartodb/dashboard/main_view_static.js

juanignaciosl commented:
Issues:
- Setting `viewer` flag should display "viewer" tag at profile page. In addition, it will probably lead to a warning message about quotas. You can check that by setting "viewer" in superadmin for any user (it's not needed that the user is inside an organization).
- Setting the avatar doesn't work. Should it work? There's a comment at the PR description but I'm not sure.
- Header should be a link to the dashboard and display "Configuration", but it won't.

Differences found during acceptance:
- Clicking "save" without changes won't trigger the submission (but has no feedback). Previous version sent it and had green feedback. Maybe not perfect but it feels like an improvement.

@matallo @elenatorro I'm not sure whether these are actual issues or scope limits. If this is known, I think that it should be ready to deploy. If it's not, move it back to development.

matallo commented:
on it!

matallo commented:
- shame on me, I hadn't taken `viewer` into account, fixed that (also extended the tests)
- fixed the avatar save (deployed and tested in ded13)
- fixed the header breadcrumb (and extended the tests)
- I also made the form always submit (even if there's no change in the model). The response is a `403` now as there are no new fields populated, though but the implementation is closer to the original design

finally, I also spotted the `upgrade_message_view` which previously pulled from `window.upgrade_url` and that is not acceptable anymore, just created a new view (and the corresponding tests

I'm moving this to CR again for the new view, and asking @elenatorro to jump in before acceptance

elenatorro commented:
I'll check it after @matallo fixes a small issue between feature flags 👍

matallo commented:
nevermind, just realized I forgot to restart the server, thus I had an outdated assets version, good to go!

juanignaciosl commented:
Differences:
- If I run out of quota (just like being a viewer but having datasets) I miss this header message:

![captura de pantalla 2017-10-09 a las 9 08 21](https://user-images.githubusercontent.com/932968/31327780-9e9d8bbc-acd1-11e7-9122-6ea28cbebe58.png)

It's likely out of scope of this PR, but please confirm.

- Minor: changing the user name and surname doesn't change the breadcrumb at header until page is reloaded. It's a minor detail, just let me know if you prefer ignoring it for the moment.
- Sending the form without changes triggers a 403. It's caused by the emptiness of `user` attribute and the server side check:

https://github.com/CartoDB/cartodb/blob/fd0f3c4183e998b8cee7d45278b762386f72de3b/app/controllers/carto/api/users_controller.rb#L62

I'm not sure about returning forbidden in this case. I'd skip processing (doing a "quick exit") but change the response be a normal 200 with content. What do you think, @mayoral, @matallo ?


All my comments are minor. @matallo , take a look and decide whether you want to merge and ignore them or address them at future PRs, or you prefer changing them in this PR.

matallo commented:
on it

matallo commented:
strange, I do have the quota header, actually that's what I was referring to in here:

> finally, I also spotted the upgrade_message_view which previously pulled from window.upgrade_url and that is not acceptable anymore, just created a new view (and the corresponding tests

the only difference is the free trial eligible, which my user isn't

![screen shot 2017-10-09 at 11 10 00](https://user-images.githubusercontent.com/36676/31332003-fe74eaea-ace4-11e7-9979-a2690fa2dde5.png)


re: the othe points

- I spotted the user name remaining until refresh, I'd leave to a separate PR (and even could do it as part of response)
- agree to change the response to a 200 when no attribute is being sent, currently on top of not being consistent with previous behaviour it says "Could not save profile, please try again." which isn't actually true. We should either change the message returned or the response code

juanignaciosl commented:
@matallo ok, let me know if you need anything. Speak to @mayoral or any other backender about the response code change if needed so we release this without more delays :)

matallo commented:
it took me fewer time adding the binding than creating a new issue, the header now updates on model save 🙌 

@mayoral is with the response change

juanignaciosl commented:
[I said this](https://github.com/CartoDB/cartodb/pull/12704#issuecomment-335083361):

> ... but change the response be a normal 200 *with content*

But this has been [the change](https://github.com/CartoDB/cartodb/pull/12704/commits/b7facabcccfb2f872d36533d2a382214b25f0e6d#diff-69e5663eb7a8349943072af4865d7b59R629: (b7facabcccfb2f872d36533d2a382214b25f0e6d):

`return(head 200) unless attributes.present?`

This doesn't return the _content_ but a response without body. In addition, after saving without changes this error is displayed, probably because of the empty body: "Could not save profile, please try again.". @mayoral please make this return the same when something changed than when it didn't. To Development (moving only one of the three issues, blocking the others at acceptance).

juanignaciosl commented:
Ready for production 🚀  Impressive mimicking work, disabling the FF doesn't show any difference :D (and I'm checking with console tools that I'm using the new one ;-) ).

FYI the quota limit header is displayed at profile page but not at dashboard one, maybe you want to address it.

I'll merge master and update package.json version now.
